torch/export/dynamic_shapes.py

```diff
@@ -371,6 +372,11 @@ def tree_zip(combined_args, dynamic_shapes):
                 )
             for k, shape in dynamic_shapes.items():
                 yield from tree_zip(combined_args[k], shape)
+        elif dataclasses.is_dataclass(combined_args) and isinstance(dynamic_shapes, (tuple, list)):
```

Comment:
I would prefer that we do this not just for dataclasses but all classes that are registered with pytree. In other words, check that `type(combined_args)` is in `SUPPORTED_NODES` and assert that `dynamic_shapes` is `list`, with an error message explaining that for user-registered classes we require you to match the flattened structure. 

Then cut the case below (that checks whether `type(dynamic_shapes) == type(combined_args)`) because that will not be possible in general and is therefore confusing.

Finally, add a test for a regular class (or better, just use KJT in the test).